homeassistant/components/subaru/config_flow.py

```diff
@@ -103,13 +108,60 @@ async def validate_login_creds(self, data):
             device_name=device_name,
             country=data[CONF_COUNTRY],
         )
-        _LOGGER.debug(
-            "Setting up first time connection to Subaru API.  This may take up to 20 seconds"
-        )
+        _LOGGER.debug("Setting up first time connection to Subaru API")
         if await self.controller.connect():
-            _LOGGER.debug("Successfully authenticated and authorized with Subaru API")
+            _LOGGER.debug("Successfully authenticated with Subaru API")
             self.config_data.update(data)
 
+    async def async_step_two_factor(self, user_input=None):
+        """Select contact method and request 2FA code from Subaru."""
+        error = None
+        if user_input:
+            # self.controller.contact_methods is a dict:
+            # {"phone":"555-555-5555", "userName":"my@email.com"}
+            selected_method = next(
+                k
+                for k, v in self.controller.contact_methods.items()
+                if v == user_input[CONF_CONTACT_METHOD]
+            )
+            if await self.controller.request_auth_code(selected_method):
+                return await self.async_step_two_factor_validate()
+            raise SubaruException("2FA request failed")
```

Comment:
Instead of raising, should we abort the flow with a defined error message?

Reply:
Yes, good catch, thank you. Fixed.